rust-peer/src/main.rs

```diff
@@ -364,3 +428,91 @@ async fn read_or_create_identity(path: &Path) -> Result<identity::Keypair> {
 
     Ok(identity)
 }
+
+// Simple file exchange protocol
+
+#[derive(Debug, Clone)]
+struct FileExchangeProtocol();
+#[derive(Clone)]
+struct FileExchangeCodec();
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct FileRequest(String);
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FileResponse(Vec<u8>);
+
+impl ProtocolName for FileExchangeProtocol {
+    fn protocol_name(&self) -> &[u8] {
+        "/universal-connectivity-file/1".as_bytes()
+    }
+}
+
+#[async_trait]
+impl request_response::Codec for FileExchangeCodec {
+    type Protocol = FileExchangeProtocol;
+    type Request = FileRequest;
+    type Response = FileResponse;
+
+    async fn read_request<T>(
+        &mut self,
+        _: &FileExchangeProtocol,
+        io: &mut T,
+    ) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        let vec = read_length_prefixed(io, 1_000_000).await?;
+
+        if vec.is_empty() {
+            return Err(io::ErrorKind::UnexpectedEof.into());
+        }
+
+        Ok(FileRequest(String::from_utf8(vec).unwrap()))
```

Comment:
If all we read is a string, do we really need to length-prefix it? It would be a cleaner protocol to just send the filename, no?

Something like:

```rust
let mut filename = String::new();
io.take(REQUEST_SIZE_MAXIMUM).read_to_string(&mut filename).await?;
```